[WIP] Add --not-required to list command.

nvdv opened the pull request:
This is an implementation of #876. Let me know what you think and I will update tests and other stuff. I am not sure how new option should be combined with existing options though.
Thanks!

pfmoore commented on pip/commands/list.py:
```diff
@@ -91,6 +91,13 @@ def __init__(self, *args, **kw):
+            '--leaves',
+            action='store_true',
+            dest='leaves',
+            help="List packages that are not dependencies of installed packages.",
```
This doesn't agree with the code, which reports all packages that have no dependencies themselves.

I'm not actually sure what the benefit is of only reporting packages with no dependencies. The original issue seemed to want the set of packages that, if you installed them, would replicate everything you had - but that's packages which aren't depended on by anything else, rather than what's implemented here.

nvdv replied:
Perhaps I misunderstood the requirements and I definitely am not familiar with all underlying technology behind the `pip`, but let's consider an example. 
I have six packages: 1, 2, 3, 4, 5, 6 with following dependencies:
1 -> []
2 -> []
3 -> []
4 -> [1, 2, 3]
5 -> [2, 3]
6 -> []
Packages with no dependencies are [1, 2, 3, 6] and packages which aren't depended on by anything else are [1, 2, 3, 6].

pfmoore replied:
Surely the packages that aren't depended on by anything else are [4, 5, 6]? And that's what I thought was the intent, as if I did ```pip install 4 5 6``` I'd get 1, 2 and 3 pulled in as dependencies, which (if I understand the explanation of what's needed from #876) is what was wanted.

Having said this, I can't *personally* think of a case where I'd want this (except maybe as the nearest I can get to what I actually care about, which is "give me all the packages I installed explicitly, but leave out any that were only pulled in as dependencies", which pip can't tell you as it doesn't retain that information)

nvdv replied:
> Surely the packages that aren't depended on by anything else are [4, 5, 6]? And that's what I thought was the intent, as if I did pip install 4 5 6 I'd get 1, 2 and 3 pulled in as dependencies, which (if I understand the explanation of what's needed from #876) is what was wanted.

OK, now I see it, thanks. I will fix the code.

> Having said this, I can't personally think of a case where I'd want this (except maybe as the nearest I can get to what I actually care about, which is "give me all the packages I installed explicitly, but leave out any that were only pulled in as dependencies", which pip can't tell you as it doesn't retain that information)

The use case I have in mind is "list of packages you can delete without breaking other packages."

pfmoore replied:
OK, cool. BTW, "leaves" doesn't really feel like the right name for "packages that are not dependencies of installed packages". The term "leaves", to me, means "packages with no dependencies", which is what you implemented. So maybe as well as fixing the implementation, this option should have a more intuitive name?

nvdv replied:
> OK, cool. BTW, "leaves" doesn't really feel like the right name for "packages that are not dependencies of installed packages". The term "leaves", to me, means "packages with no dependencies", which is what you implemented. 

I have Bash script that does exactly what you mentioned (and that's the reason I created this PR) and for some reason I didn't implement it in the first commit. I guess I should do less things in parallel 😄 
I have fixed the code and changed the name of the option. Let me know what you think.
BTW, thanks for the quick responses!

pfmoore commented on pip/commands/list.py:
```diff
@@ -146,15 +153,20 @@ def run(self, options, args):
                 "Options --outdated and --uptodate cannot be combined.")
 
+        # TODO (nvdv): Check options combinations.
+
```
As noted, checks that incompatible combinations of arguments aren't given need to be added.

In theory, I guess ```--nodeps``` could be used with many of the other options, but I'd be inclined to say YAGNI for now.

xavfernandez replied:
I'd say we either implement the option completely (in combination with other options) or `pip list` should emit a warning saying explicitely (or crash like for --outdated/--uptodate) when the option is ignored.
(it does not seem hard to implement it completely though).

nvdv replied:
Now `self.get_no_deps_only` acts as final filter for other options. Let me know If I missed somethin.

xavfernandez commented on pip/commands/list.py:
```diff
@@ -91,6 +91,13 @@ def __init__(self, *args, **kw):
         )
 
+        cmd_opts.add_option(
+            '--nodeps',
```
bikeshed: wondering if `--no-deps-only` would not be clearer ?

nvdv replied:
Done.

nvdv commented on pip/commands/list.py:
```diff
@@ -169,6 +181,15 @@ def get_uptodate(self, packages, options):
+        for dist in installed_packages:
+            all_dependencies.extend(dist.requires())
+        dep_keys = set(dep.key for dep in all_dependencies)
+        return set(pkg for pkg in installed_packages if pkg.key not in dep_keys)
```
IIUC `installed_packages` and `deps_keys` contain objects of different types. `dist.requires()` returns `Requirement` objects and `dist` has other type.

pfmoore replied:
Whoops, my bad, you are perfectly correct.

xavfernandez replied:
nit: you don't need a temporary list `all_dependencies`, I think you could do:

    dependencies_keys = set()
    for dist in installed_packages:
        dependencies_keys.update(dep.key for dep in dist.requires())
    return set(pkg for pkg in installed_packages if pkg.key not in dependencies_keys)

nvdv replied:
Good catch! 
Fixed.

xavfernandez commented:
It looks good to me, a small changelog entry plus a small example in https://github.com/pypa/pip/blob/master/docs/reference/pip_list.rst and it should be good to go :+1:

nvdv commented:
I've just added the small example to docs. If it looks good to you, I'll update all necessary tests.

nvdv commented:
any news here?

xavfernandez commented:
@nvdv it looks good except a few pep8 failures (lines too long).
It just needs pep8 fix, a changelog and a few tests :+1:

nvdv commented:
I have fixed PEP8 errors and added few lines to CHANGES.txt. 
IIUC to test this I need to add a few test cases to `tests/functional/test_list.py`.  
Let's consider this code:

nvdv commented:
I have fixed PEP8 errors and added few lines to CHANGES.txt.
IIUC to test this I need to add a few test cases to `tests/functional/test_list.py`.

Let's consider this code
```python 
def test_no_deps_only_flag(script, data):
    script.pip(
        'install', '-f', data.find_links, '--no-index', 'requires_simple_extra'
    )
    result = script.pip('list', '--no-deps-only')
    ...
```
When I run this code, `installed_packages` variable inside `get_nodeps_only` equals to `['pip', 'setuptools']`. `requires_simple_extra` is not listed there and as a result test does not work. I've tried some other test packages with requirements, but results were the same. 
It seems that test environment needs to be adjusted for this test case. 
Do you have any ideas?

xavfernandez commented:
@nvdv Well, is `requires_simple_extra` successfully installed in your test ?

nvdv commented:
> Well, is requires_simple_extra successfully installed in your test ?

yep, it's present in the output of `script.pip('list')`

xavfernandez commented on pip/commands/list.py:
```diff
@@ -169,6 +182,15 @@ def get_uptodate(self, packages, options):
 
+    def get_nodeps_only(self, packages, options):
+        installed_pkgs = [
+            dist for dist in self.iter_packages_latest_infos(packages, options)
```
I'm on mobile so hard to check but I think you should not be calling `iter_packages_latest_infos` here.
You should only be filtering based on dependencies.

xavfernandez replied:
Yup, you should only iter on `packages`

nvdv replied:
Done. Thanks!

xavfernandez commented:
@nvdv Thanks for your work.
I'm ready to merge it, I'm just not a fan of the option name.
Wondering if `--not-required` (as list the packages that are not required by other packages) would not be clearer ?

nvdv commented:
> @nvdv Thanks for your work.

Sure. Thanks for quick responses.

> Wondering if --not-required (as list the packages that are not required by other packages) would not be clearer ?

Done.

xavfernandez commented:
Thank you :)

nvdv commented:
Great, thanks!
btw, you can close #876 as well

xavfernandez commented:
Closed :+1:
